### PR TITLE
Remove Docker mentions in Resource Usage Monitoring

### LINKED ISSUE #30976: Action on Docker mentions in Tasks > Monitoring Logging and Debugging
Umbrella issue #30771 
Fixes issue #30929 
---

File: /tasks/debug-application-cluster/resource-usage-monitoring.md
URL: https://k8s.io/docs/tasks/debug-application-cluster/resource-usage-monitoring

Line 44 mentions how the kubelet pulls info for Docker integration for metrics-server.

Action: Review to see if it works as intended in a world without dockershim.

## Done

~~File: /tasks/debug-application-cluster/crictl.md~~
~~URL: https://k8s.io/docs/tasks/debug-application-cluster/crictl~~

~~The entire page mentions and implements dockershim. Also includes guides for mapping docker cli commands to crictl.~~

~~Action: Rewrite the entire page without dockershim and docker CLI~~

---

~~File: /tasks/debug-application-cluster/debug-application.md~~
~~URL: https://k8s.io/docs/tasks/debug-application-cluster/debug-application~~

~~Line 61 mentions doing `docker pull <image>`.~~

~~Action: Provide an alternative or remove the line.~~

---

~~File: /tasks/debug-application-cluster/debug-pod-replication-controller.md~~
~~URL: https://k8s.io/docs/tasks/debug-application-cluster/debug-pod-replication-controller~~

~~Line 89 mentions doing `docker pull <image>`. ~~

~~Action: Provide an alternative or remove the line.~~

---

File: /tasks/debug-application-cluster/debug-application-introspection.md
URL: https://k8s.io/docs/tasks/debug-application-cluster/debug-application-introspection

Line 106 and other output examples use docker as the runtime. 

Action: Reproduce the output samples on a cluster not running Docker and update the samples. 

---
kubernetes/website/content/en/docs/tasks/debug-application-cluster/troubleshooting.md
File: /tasks/debug-application-cluster/troubleshooting.md
URL: https://k8s.io/docs/tasks/debug-application-cluster/troubleshooting

Line 103 mentions specifying the Docker version when filing a Github issue. 

Action: Say instead to mention the runtime and version being used.

---

File: /tasks/debug-application-cluster/debug-service.md
URL: https://k8s.io/docs/tasks/debug-application-cluster/debug-service

Line 679 has `--runtime-cgroups=/docker-daemon`. 

Action: Reproduce using a different runtime and update.

/sig docs
/language en

### PR DESCRIPTION
**Pull request by shannonxtreme**
Fixes #30976

Affected page: https://kubernetes.io/docs/tasks/debug-application-cluster/resource-usage-monitoring

Previously, a sentence said that "the kubelet fetches metrics from cAdvisor for *the legacy Docker integration*. Updated to encompass all cgroup based container runtimes. 

/sig docs
/language en
/milestone 1.24

### CODE REVIEW

**Review by sftim — COMMENTED** [review]
If cAdvisor is only used with the dockershim, we'd need to reword further.

File: content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md [hunk 1]
```diff
@@ -41,7 +41,7 @@ the nodes, managing the pods and containers running on a machine. The kubelet
 translates each pod into its constituent containers and fetches individual 
 container usage statistics from the container runtime through the container 
 runtime interface. The kubelet fetches this information from the integrated 
-cAdvisor for the legacy Docker integration.  It then exposes the aggregated pod 
+cAdvisor for cgroup-based container runtimes.  It then exposes the aggregated pod
```
[the remark this answers is not in the document]
Comment (shannonxtreme) [reply content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md:44]: Your suggestion LGTM Tim. I reworded this based on [my question in #sig-node](https://kubernetes.slack.com/archives/C0BP8PW9G/p1645657340951729) if that helps

[the remark this answers is not in the document]
Comment (reylejano) [reply content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md:44]: There is a KEP scheduled to move to beta in 1.24 to not use cAdvisor for metric collection and to use the CRI implementation
Perhaps this can be part of the documentation change for 1.24 for this KEP
[SIG Node KEP 2371 cAdvisor-less, CRI-full Container and Pod Stats](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2371-cri-pod-container-stats)
/sig node

cc @nate-double-u

[the remark this answers is not in the document]
Comment (shannonxtreme) [reply content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md:44]: I like that idea Rey. So it's expanding the CRI summary API to rely on the CRI implementation for Pod and container stats instead of relying on cAdvisor? Cool! What would I need to do here? Change the destination branch to dev-1.24?

File: content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md [hunk 2]
```diff
@@ -40,8 +40,11 @@ memory usage. The kubelet acts as a bridge between the Kubernetes master and
 the nodes, managing the pods and containers running on a machine. The kubelet 
 translates each pod into its constituent containers and fetches individual 
 container usage statistics from the container runtime through the container 
-runtime interface. The kubelet fetches this information from the integrated 
-cAdvisor for the legacy Docker integration.  It then exposes the aggregated pod 
+runtime interface. If you use a container runtime that uses Linux cgroups and
+namespacesto implement containers, and the container runtime does not publish
```
Comment (sftim) [inline content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md:44]:
```suggestion
namespaces to implement containers, and the container runtime does not publish
```

File: content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shannonxtreme) [reply content/en/docs/tasks/debug-application-cluster/resource-usage-monitoring.md]: @reylejano if you want we can discuss this on Slack and I'm happy to get involved in the docs changes for this KEP for 1.24. LMK :)

**Comment by sftim** [comment]
@kubernetes/sig-node-pr-reviews does this look technically correct?

**Comment by dims** [comment]
cc @bobbypage

File: content/en/docs/tasks/debug/debug-cluster/resource-usage-monitoring.md [hunk 3]
```diff
@@ -41,8 +41,11 @@ memory usage. The kubelet acts as a bridge between the Kubernetes master and
 the nodes, managing the pods and containers running on a machine. The kubelet 
 translates each pod into its constituent containers and fetches individual 
 container usage statistics from the container runtime through the container 
-runtime interface. The kubelet fetches this information from the integrated 
-cAdvisor for the legacy Docker integration.  It then exposes the aggregated pod
```
Comment (bobbypage) [inline content/en/docs/tasks/debug/debug-cluster/resource-usage-monitoring.md:45]: I would reword this to something like:

"The kubelet fetches this information by combining usage statics from from the container runtime as well as cAdvisor."

**Comment by bobbypage** [comment]
> @kubernetes/sig-node-pr-reviews does this look technically correct?

It's not 100% correct, kubelet still uses cAdvisor to obtain some usage metrics even on a CRI runtime and when the CRI runtime exposes usage stats. Currently kubelet fetches stats from both the CRI runtime as well as cAdvisor and combines the stats together to expose them via the summary API. This is because not all stats needed for summary API are exposed by CRI. 

https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2371-cri-pod-container-stats is long term KEP that aims to ensure that we can only rely on CRI and eliminate needing cAdvisor for pod and container level stats, but the KEP is currently alpha and still WIP.

File: content/en/docs/tasks/debug/debug-cluster/resource-usage-monitoring.md (hunk 3 above)
Comment (shannonxtreme) [reply content/en/docs/tasks/debug/debug-cluster/resource-usage-monitoring.md:45]: This is the current wording: 

If you use a container runtime that uses Linux cgroups and
namespaces to implement containers, and the container runtime does not publish
usage statistics, then the kubelet can look up those statistics directly
(using code from [cAdvisor](https://github.com/google/cadvisor)).
No matter how those statistics arrive, the kubelet then exposes the aggregated pod
resource usage statistics through the metrics-server Resource Metrics API.


Is that more correct?

**Comment by sftim** [comment]
If I use something like a partitioning hypervisor and a runtime that speaks CRI, does cAdvisor get involved in that case (if not - what wording can we use to make the situation clear)?

We don't have to get things 100% right (I'd settle for less wrong - and removing an implication that Docker Engine is part of the story ought to count). However, if someone is considering an unusual but CRI-compatible way to run Pods, it'd be nice to land a PR that lets then work out the actual story.

**Comment by zacharysarah** [comment]
I see the PR author [responding to feedback](https://github.com/kubernetes/website/pull/32069/commits/db056570a9f589b2a4f2a56ff11064ae3d6688fd), so I'd like to move this PR forward. I think their amendments seem reasonable based on the feedback given. 

I'm going to approve and LGTM this PR, and put an explicit 48-hour `/hold` in place to give previous reviewers a chance to weigh in. @dims @bobbypage @sftim 👋🏻 Please feel free to `/hold cancel` in less than 48 hours if you're OK with this PR merging as is. 🙇🏻 

/lgtm
/approve
/hold

**Comment by sftim** [comment]
Thanks @zacharysarah

**Comment by sftim** [comment]
All the concerns I saw were on the SIG Node / technical accuracy side, so I'll let that hold stand a while longer.

**Comment by sftim** [comment]
Also, @shannonxtreme, if you'd like a Zoom call or similar to go over the nuances here, we can try to plan that in. That can be either before or after these changes merge.

**Comment by shannonxtreme** [comment]
Yeah @sftim I'm happy to do a call to step through the nitty-gritty so that we can maybe do a good-first-issue to update the doc after this PR is merged?

**Comment by zacharysarah** [comment]
Hi folks--48+ hours have passed without any blocking reviews, so I think it's time to lift the hold.

/hold cancel
